boot/flags.go

```diff
@@ -0,0 +1,111 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2021 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"fmt"
+	"regexp"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader"
+)
+
+func blForDev(dev Device) (bootloader.Bootloader, error) {
+	opts := &bootloader.Options{}
+	dir := ""
+	if dev.RunMode() {
+		opts.Role = bootloader.RoleRunMode
+	} else {
+		opts.Role = bootloader.RoleRecovery
+		// meh this isn't being used in the initramfs but it's probably fine
+		dir = InitramfsUbuntuSeedDir
+	}
+
+	return bootloader.Find(dir, opts)
+}
+
+// NextBootFlags returns the set of boot flags for the current active boot and
+// possibly for the next boot. By default, the flags should only be used on one
+// boot ever after being set and the system being rebooted with the flags
+// cleared by snapd in userspace when that boot happens. The mode parameter is
+// necessary to determine the current active bootloader. Only to be used on
+// UC20+ on systems with recovery systems.
+func NextBootFlags(dev Device) ([]string, error) {
+	if !dev.HasModeenv() {
+		return nil, fmt.Errorf("cannot get boot flags on non-UC20 device")
+	}
+
+	bl, err := blForDev(dev)
+	if err != nil {
+		return nil, err
+	}
+
+	m, err := bl.GetBootVars("snapd_next_boot_flags")
```

Comment:
I forgot, didn't we discuss/consider passing this over the kernel cmdline (so that it would need to be sealed against for mode != install)

Reply:
Hmm, here's what I have in my notes about this, I have it written that `snapd_next_boot_flags` should be set for the bootenv by ubuntu-image for install mode (which suggest that it is not on the kernel command line for install mode), but also that the flag should be auto-clearing if nothing else happens (i.e. hooks deciding what to do during install mode). Then I have it for the "run mode in the factory" feature, we could put it on the kernel command line, but I don't have more details about doing that here. 

I think whether this goes on the kernel command line for run mode, etc. depends on how we want to implement using this flag for non-install modes. If this flag is set for run mode on a first boot and we do privileged things that an attacker couldn't do otherwise with disk access, then having this flag sealed in the kernel command line makes sense. An example of where this doesn't make sense is that if at the end of this factory run mode we wipe the data partition (or maybe we mount a tmpfs overlay on top of the real ubuntu-data as it exists at the end of install mode), then I don't think it matters, since if an attacker adds this flag when it shouldn't be there, they just end up wiping the system, but since they had disk access they could have just wiped the disk anyways and gotten the same effect. If however we are running some other snap services or something in this "factory run mode" that might make device agent snaps do something privileged like create a WiFi network or do something with hardware that's expected to be in the factory for testing like a USB drive or something, then we definitely would want to seal this flag to the kernel command line to protect against an attacker causing this to happen outside of the factory.

However, we have to consider the difficulty of doing this, since I think resealing to include this on the kernel command line and re-writing grub.cfg for run mode to pick up this flag means that at least for now I think it is not doable, since we haven't really finished the work around modifying the kernel command line from a gadget, and I think that including this flag on the kernel command line would require that work (and note that we need to go the other way too, to drop this flag from the kernel command line at the end too).

Reply:
yes, it would need grub.cfg changes but not particularly specific to anything else (but again if we do have grub.cfg changes we want to bundle them together). Anyway let's chat a bit tomorrow on this as agreed.